src/lib.rs

```diff
@@ -103,12 +103,133 @@ impl TryFrom<u8> for HashAlgorithm {
         match value {
             // Naturally, these need to match the enum declaration
             1 => Ok(Self::MurmurHash3),
-            2 => Ok(Self::Sha256),
+            2 => Ok(Self::Sha256l32),
+            3 => Ok(Self::Sha256),
             _ => Err(()),
         }
     }
 }
 
+trait CascadeIndexGenerator {
+    fn next_index(&mut self, range: usize) -> usize;
+    fn next_layer(&mut self);
+}
+
+struct MurmurHash3IndexGenerator<'a> {
+    key: &'a [u8],
+    counter: u32,
+    depth: u32,
+}
+
+impl<'a> MurmurHash3IndexGenerator<'a> {
+    fn new(key: &'a [u8]) -> Self {
+        MurmurHash3IndexGenerator {
+            key: key,
+            counter: 0,
+            depth: 1,
+        }
+    }
+}
+
+impl<'a> CascadeIndexGenerator for MurmurHash3IndexGenerator<'a> {
+    fn next_index(&mut self, range: usize) -> usize {
+        let hash_seed = (self.counter << 16) + self.depth as u32;
+        self.counter += 1;
+        let index = murmurhash3_x86_32(self.key, hash_seed) as usize;
+        index % range
+    }
+    fn next_layer(&mut self) {
+        self.counter = 0;
+        self.depth += 1
+    }
+}
+
+struct SHA256l32IndexGenerator<'a> {
+    salt: &'a [u8],
+    key: &'a [u8],
+    counter: u32,
+    depth: u8,
+}
+
+impl<'a> SHA256l32IndexGenerator<'a> {
+    fn new(salt: &'a [u8], key: &'a [u8]) -> Self {
+        SHA256l32IndexGenerator {
+            salt: salt,
```

Comment:
When a field has the same name as the variable being assigned to it, you can omit the field name, but it's not a big deal:
```
SHA256l32IndexGenerator {
    salt,
    key,
```